### PR TITLE
Add SDC/CloudAPI provider

### PR DESCRIPTION
**Pull request by MerlinDMC**
This will add a basic provider for SDC/CloudAPI.
The official golang client libraries provided by joyent are used for api calls. (https://github.com/joyent/gosdc and https://github.com/joyent/gocommon)

It currently supports instance create/update/destroy and reads configuration from the same ENV variables as the smartdc cli utilities.

Updating the machine package is working but highly depends on the source and target packages which under some conditions will be blocked by CloudAPI.
For example: Upgrading from a `Standard` to a `High I/O` package will not work.

The acceptance tests are configured to run in the AMS-1 datacenter.

### CODE REVIEW

**Comment by catsby** [comment]
Hey @MerlinDMC! Thanks for the contribution here. Would it be possible to also get documentation added for this new Provider and these new resources? 

Instructions on running the website locally are in the README in the website folder 

- https://github.com/hashicorp/terraform/tree/master/website

You can then likely take an existing provider like the [AWS Provider][1] and use it as a framework for documenting, just rip most things out. 

Adding documentation will help us evaluate and merge your request quicker, as well as help the community use your new Provider. 

Thanks!

[1]: https://github.com/hashicorp/terraform/tree/master/website/source/docs/providers/aws

**Comment by MerlinDMC** [comment]
@catsby sure ... will take some time since I drown in dayjob work probably till the weekend.
But it would be good to get a native speaker prove reading that doc once It's commited ;)

**Comment by bbhoss** [comment]
Been waiting on this for a long time, good work!

**Comment by MerlinDMC** [comment]
Would be nice if someone could have a look at the docs and make annotations for each `WTF?` moment

File: website/source/docs/providers/sdc/index.html.markdown [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+used.
+
+Use the navigation to the left to read about the available resources.
+
+## Example Usage
+
+```
+# Configure the SDC Provider
+provider "sdc" {
+    sdc_key_name = "${var.dc_key_name"
```
Comment (bbhoss) [inline website/source/docs/providers/sdc/index.html.markdown]: Typo?

File: website/source/docs/providers/sdc/r/instance.html.markdown [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+* `package` - (Required) The instance package.
+* `network` - (Required) The list of networks to attach to the resource.
+  See [Networks](#networks) below for details.
+* `metadata` - (Optional) A mapping of metadata to assign to the resource.
+* `tags` - (Optional) A mapping of tags to assign to the resource.
+
+<a id="networks"></a>
+## Networks
+
+The networks block controls what network should be attached to the given
```
Comment (bbhoss) [inline website/source/docs/providers/sdc/r/instance.html.markdown]: How about
```markdown
The `network` block controls which networks should be attached to an instance.
```

File: website/source/layouts/docs.erb [hunk 3]
```diff
@@ -181,6 +181,10 @@
 					<a href="/docs/providers/rundeck/index.html">Rundeck</a>
 					</li>
 
+					<li<%= sidebar_current("docs-providers-sdc") %>>
+            <a href="/docs/providers/sdc/index.html">SDC/CloudAPI</a>
```
Comment (bbhoss) [inline website/source/layouts/docs.erb]: Looks like the spacing is off here

File: website/source/layouts/sdc.erb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<% wrap_layout :inner do %>
+	<% content_for :sidebar do %>
+		<div class="docs-sidebar hidden-print affix-top" role="complementary">
```
Comment (bbhoss) [inline website/source/layouts/sdc.erb]: ditto. Tab/space mix maybe?

**Comment by bbhoss** [comment]
Provided some quick comments. All in all looks pretty good.

**Comment by MerlinDMC** [comment]
@catsby Is there anything else that is missing now that the docs are added?

**Comment by benjamin-bergia** [comment]
Can't wait to see this provider included :)

**Comment by jen20** [comment]
Hi @MerlinDMC et al! I apologise for the amount of time it took for us to get to reviewing this. It turns out that since this was opened, however, Joyent have built an "official" provider for Terraform for use with the SDC APIs - it's available here: https://github.com/joyent/triton-terraform. We have discussed merging it into Terraform core as a builtin, which is likely to happen when it has proven stable (my personal use of it is fairly positive so far).

Consequently I'm going to close this PR 'in favour' of the Joyent repo to avoid having multiple different SDC providers floating around. Many thanks for the amount of work clearly went into this - even from a personal perspective it is great to see people supporting Illumos and related technologies.

Also FWIW as of 0.6.9 of Terraform, there are builds which run on Solaris-derived operating systems including SmartOS.

**Pull request closed without merging**